sympy/solvers/solveset.py

```diff
@@ -239,6 +239,22 @@ def _invert_real(f, g_ys, symbol):
         if g is not S.One:
             return _invert_real(h, imageset(Lambda(n, n/g), g_ys), symbol)
 
+        if g_ys == FiniteSet(S.Zero):
+            res = S.EmptySet
+            test = True
+            for f1 in f.args:
+                (a, b) = _invert_real(f1, g_ys, symbol)
+                if a == symbol and isinstance(b, FiniteSet):
+                    if all(checksol(f, symbol, sol) for sol in b):
+                        res = res.union(b)
```

Comment:
It would be better not to use `checksol` here but also I'm not sure how the logic can be formally correct. Why do we check with `all`? What would happen if some of the solutions are good and some are not?

I would have thought that it would be best to have something here to split the numerator and denominator. If you want to invert `f(x)/g(x) = a` then (provided `g(x)!=0`) that's the same as inverting `f(x) - a*g(x) = 0`.

Also does the same issue not arise with the rhs being nonzero like this:
```python
In [16]: solveset(sqrt((x-3)/x)-Rational(1, 2), x, Reals)
Out[16]: ∅
```
I'm not sure where the logic falls through here but at least it should be a ConditionSet rather than simply an empty set.

Reply:
I used `checksol` to eliminate non correct solutions like `x=0` in `log(1+x)/x`, which produced a failure in tests. Is there any obstruction to the use of `checksol`?

You right. The check with `all` is not good. I will change it including the possibility that some solutions are good and some not.

I imagine that splitting numerator and denominator works in the case. But `invert_real((x-3)*(x+5), 0, x)` returns `((x - 3)*(x + 5), {0})` and in this case `invert_real` should not be affected. Also `solveset` in this case gives the wrong result.

I do not understand your last comment. `x=4` is a solution. The output seems wrong.

Reply:
Since `checksol` can be very expensive it should be used sparingly. There should be a simpler way of eliminating `x=0` from `log(1+x)/x` i.e. we want roots of the numerator that are not roots of the denominator.

> But `invert_real((x-3)*(x+5), 0, x)` returns `((x - 3)*(x + 5), {0})`

Well `solveset` can solve this so where in `solveset` is this handled? Maybe the case considered here should be handled there as well.

> I do not understand your last comment. `x=4` is a solution. The output seems wrong.

Yes, the output is wrong and the changes here don't fix it even though it is the same function to be inverted. That's because the solution here doesn't really invert the function: it's a special case for when the rhs is zero.

Reply:
What do you think about the code?
```
        if g_ys == FiniteSet(S.Zero):
            res = _solve_as_rational(f, symbol, S.Reals)
            return (symbol, res)
```
The function `_solve_as_rational()` works in the spirit of your suggestion.

> Yes, the output is wrong and the changes here don't fix it even though it is the same function to be inverted. That's because the solution here doesn't really invert the function: it's a special case for when the rhs is zero.

I take a look of the case for understanding where the logic does not work.

Reply:
I'm not sure that this really belongs in `invert` rather than somewhere else. It doesn't look like `invert` is intended to handle anything nontrivial that can't just be inverted one function at a time e.g.:
```python
In [20]: invert_real(x**2 + x, 0, x)
Out[20]: 
⎛ 2         ⎞
⎝x  + x, {0}⎠
```

Reply:
> I'm not sure that this really belongs in `invert` rather than somewhere else. It doesn't look like `invert` is intended to handle anything nontrivial that can't just be inverted one function at a time e.g.:

Solving `f(g(x))=0` can be simplified when `invert_real` becomes more powerful, as we proceed to solving `g(x)=f_inv(0)` instead. In your example `g(x)=x` which indeed doesn't show benefit. But what if `g(x)` is much more complicated?

Reply:
After some debug, I am convinced that the issue lies in a return of `_invert_real()`, in which a set were intersected with a domain for base positive.
I think that this is fundamental only if the inverse function really works.

In particular this change seems to fix also `solveset(sqrt((x-3)/x)-Rational(1, 2), x, Reals)`.